### PR TITLE
fixed errors with factorial going to 0.0 when overflowed by using f64…

### PR DESCRIPTION
**Pull request by Dev380**
…, side effects include precision loss and error handling of any overflows being improved

### CODE REVIEW

**Comment by Dev380** [comment]
I just noticed that many files seem to be modified, the important logic is in `error.rs` and `lex.rs`, the others are from using `cargo fmt`.

File: src/lex.rs [hunk 1]
```diff
@@ -122,7 +129,11 @@ lazy_static! {
 }
 
 fn factorial(n: f64) -> f64 {
-    n.signum() * (1..=n.abs() as u64).fold(1, |p, n| p * n) as f64
+    let answer = (1..=n.round() as u128).map(|u| u as f64).product::<f64>() as f64;
+    if answer == 0.0 {
+        return f64::INFINITY;
+    }
```
Comment (pickfire) [inline src/lex.rs:149]: Why is this infinity if it's zero? So 0.0 is not a valid value?

Comment (Dev380) [reply src/lex.rs:149]: I don't think 0.0 is in the domain of the discrete factorial function. It'd be nice to add support for the gamma function but that's not in the scope of this PR.

Comment (Dev380) [reply src/lex.rs:149]: However, you did make me notice that I didn't handle 0 as an input value, I think I'll just hardcode that as 1.

File: src/lex.rs [hunk 2]
```diff
@@ -32,9 +32,18 @@ impl Operator {
     }
     pub fn operate(self, x: f64, y: f64) -> Result<f64, CalcError> {
         if self.token == '/' && y == 0. {
-            Err(CalcError::Math(Math::DivideByZero))
+            return Err(CalcError::Math(Math::DivideByZero));
+        } else if (self.token == '!' && x < 0.0) || (self.token == '!' && x.fract() != 0.0) {
+            return Err(CalcError::Math(Math::OutOfBounds));
+        } else if self.token == '!' && x == 0.0 {
+            // Must return 1 manually as 0..0 which the factorial function does doesn't work AFAIK.
```
Comment (pickfire) [inline src/lex.rs]: Is `0..0` a typo?

Comment (Dev380) [reply src/lex.rs]: Ah sorry, I meant 0.0..=n where n=0.0

**Review by pickfire — APPROVED** [review]
I think should be good but not tested.

File: src/main.rs [hunk 3]
```diff
@@ -285,9 +282,6 @@ mod tests {
     }
     #[test]
     fn eval_e_times_n() {
-        assert_eq!(
-            0. as f64,
-            eval_math_expression("e0", None).unwrap()
-        );
+        assert_eq!(0. as f64, eval_math_expression("e0", None).unwrap());
```
Comment (pickfire) [inline src/main.rs]: Would it be good to have the current changes unit tested?

File: src/lex.rs [hunk 4]
```diff
@@ -32,9 +32,18 @@ impl Operator {
...
     pub fn operate(self, x: f64, y: f64) -> Result<f64, CalcError> {
         if self.token == '/' && y == 0. {
-            Err(CalcError::Math(Math::DivideByZero))
+            return Err(CalcError::Math(Math::DivideByZero));
+        } else if (self.token == '!' && x < 0.0) || (self.token == '!' && x.fract() != 0.0) {
+            return Err(CalcError::Math(Math::OutOfBounds));
+        } else if self.token == '!' && x == 0.0 {
+            // Must return 1 manually as 0..=n where n is 0.0 doesn't work AFAIK.
+            return Ok(1.0);
+        }
```
Comment (pickfire) [inline src/lex.rs:28]: I don't think having `!` at `operate` is the correct place, it should only be checked for `factorial!`. But seemed like we don't have a base to handle this. As of now can just leave a TODO.

**Comment by pickfire** [comment]
Thanks a lot for contributing! Sorry for the late reply, I have updated the master branch, can you please resolve the conflicts?

**Comment by pickfire** [comment]
The factorial issue is partially fixed with https://github.com/nerdypepper/eva/commit/7cc7775d25e66f42de1ae8921e25c95c5f9fe25e

**Review by pickfire — COMMENTED** [review]
Thanks for the quick response, but can you please add some tests to test out the two errors?

**Comment by pickfire** [comment]
I opened another pull request to fix the stuff mentioned here since author is inactive. https://github.com/nerdypepper/eva/pull/60

Thanks a lot for working on this.

**Pull request closed without merging**